first version of the well testing for physical reason

GitPaean opened the pull request:
the key difficulty is that we do not have reliable explicit information
to do the testing.

In this version, we try to obtain the explicit information by finishing
one converged solving.

GitPaean committed:
first version of the well testing for physical reason

the key difficulty is that we do not have reliable explicit information
to do the testing.

In this version, we try to obtain the explicit information by finishing
one converged solving.

GitPaean commented:
jenkins build this please

akva2 commented on opm/autodiff/MultisegmentWell_impl.hpp:
```diff
@@ -1648,7 +1648,7 @@ namespace Opm
                          const WellState& /* well_state */)
     {
-        const std::string msg = "Support of well operatability checking for mutlisegment wells is not implemented "
+        const std::string msg = "Support of well operability checking for mutlisegment wells is not implemented "
```
may as well fix the typo while at it. multisegment

akva2 commented on opm/autodiff/MultisegmentWell_impl.hpp:
```diff
@@ -1888,4 +1888,20 @@ namespace Opm
+                        const double simulation_time, const int report_step, const bool terminal_output,
+                        WellState& well_state, WellTestState& welltest_state)
+    {
+        const std::string msg = "Support of well testing for physical limits for mutlisegment wells is not "
```
and here.

GitPaean replied:
Thanks for catching them.

GitPaean commented:
jenkins build this please

atgeirr commented on opm/autodiff/MultisegmentWell.hpp:
```diff
@@ -349,6 +349,10 @@ namespace Opm
+
+        virtual void wellTestingPhysical(Simulator& simulator, const std::vector<double>& B_avg,
+                                         const double simulation_time, const int report_step, const bool terminal_output,
+                                         WellState& well_state, WellTestState& welltest_state);
```
I get a compiler warning: this function should be marked `override`, same in StandardWell.hpp.

GitPaean commented:
jenkins build this please

GitPaean committed:
not updateWellControl for non-operable wells

GitPaean commented:
jenkins build this please

GitPaean committed:
typo correction and message output adjustment for WTEST

recovering the const for updateWellStateWithTarget

there is no functional change.

atgeirr commented:
This PR can be merged, we are only waiting for a final run to complete. @GitPaean if successful, you can merge this.

GitPaean commented:
jenkins build this please

GitPaean commented:
We hold this PR for a little longer. It runs through, while some more investigation is needed before merging this.

GitPaean committed:
not shut a well because of cannot-produce-inject reason

which is a numerical reason. Removing it for now for better result.

It minght come back with more complete strategy and understanding.

GitPaean commented:
jenkins build this please

atgeirr commented:
This has been tested and found good. We'll merge this now. It's been a long journey!
